// @graphql-codegen/near-operation-file-preset@2.5.0/cjs/resolve-document-imports.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveDocumentImports = void 0;
const tslib_1 = require("tslib");
const plugin_helpers_1 = require("@graphql-codegen/plugin-helpers");
const visitor_plugin_common_1 = require("@graphql-codegen/visitor-plugin-common");
const fragment_resolver_js_1 = tslib_1.__importDefault(require("./fragment-resolver.js"));
/**
 * Transform the preset's provided documents into single-file generator sources, while resolving fragment and user-defined imports
 *
 * Resolves user provided imports and fragment imports using the `DocumentImportResolverOptions`.
 * Does not define specific plugins, but rather returns a string[] of `importStatements` for the calling plugin to make use of
 */
function resolveDocumentImports(presetOptions, schemaObject, importResolverOptions, dedupeFragments = false) {
    const resolveFragments = (0, fragment_resolver_js_1.default)(importResolverOptions, presetOptions, schemaObject, dedupeFragments);
    const { baseOutputDir, documents } = presetOptions;
    const { generateFilePath, schemaTypesSource, baseDir, typesImport } = importResolverOptions;
    return documents.map(documentFile => {
        try {
            const generatedFilePath = generateFilePath(documentFile.location);
            const importStatements = [];
            const { externalFragments, fragmentImports } = resolveFragments(generatedFilePath, documentFile.document);
            const externalFragmentsInjectedDocument = {
                ...documentFile.document,
                definitions: [...documentFile.document.definitions, ...externalFragments.map(fragment => fragment.node)],
            };
            if ((0, plugin_helpers_1.isUsingTypes)(externalFragmentsInjectedDocument, [], schemaObject)) {
                const schemaTypesImportStatement = (0, visitor_plugin_common_1.generateImportStatement)({
                    baseDir,
                    emitLegacyCommonJSImports: presetOptions.config.emitLegacyCommonJSImports,
                    importSource: (0, visitor_plugin_common_1.resolveImportSource)(schemaTypesSource),
                    baseOutputDir,
                    outputPath: generatedFilePath,
                    typesImport,
                });
                importStatements.unshift(schemaTypesImportStatement);
            }
            return {
                filename: generatedFilePath,
                documents: [documentFile],
                importStatements,
                fragmentImports,
                externalFragments,
            };
        }
        catch (e) {
            throw new Error(`Unable to validate GraphQL document! \n
         File ${documentFile.location} caused error:
         ${e.message || e.toString()}`);
        }
    });
}
exports.resolveDocumentImports = resolveDocumentImports;
